server/athenian/api/precompute/accounts.py

```diff
@@ -418,47 +404,99 @@ async def report_precompute_success():
-    num_bots = await _ensure_bot_team(account, bots, root_team_id, prefixer, sdb, mdb)
+    try:
+        root_team_id = await _ensure_root_team(account, sdb)
+    except Exception as e:
+        log.warning("root team %d: %s: %s", account, type(e).__name__, e)
+        sentry_sdk.capture_exception(e)
+        return 0, 0
+    try:
+        num_bots = await _ensure_bot_team(account, bots, root_team_id, prefixer, sdb, mdb, log)
+    except Exception as e:
+        log.warning("bots %d: %s: %s", account, type(e).__name__, e)
+        sentry_sdk.capture_exception(e)
+        num_bots = 0
+        # no return
+    if not precomputed:
+        log.info("considering account %d as brand new, creating the teams", account)
+        try:
+            _, num_teams = await copy_teams_as_needed(
+                account, meta_ids, root_team_id, sdb, mdb, cache,
+            )
+        except Exception as e:
+            log.warning("teams %d: %s: %s", account, type(e).__name__, e)
+            sentry_sdk.capture_exception(e)
+            num_teams = 0
+            # no return
+    else:
+        num_teams = 0
     return num_teams, num_bots
 
 
 async def _ensure_bot_team(
     account: int,
-    local_bots: Collection[str],
+    bots: frozenset[str] | set[str],
     root_team_id: int,
     prefixer: Prefixer,
     sdb: Database,
     mdb: Database,
+    log: logging.Logger,
 ) -> int:
-    bot_team = await get_bots_team(account, sdb)
-    if bot_team is not None:
-        return len(bot_team[Team.members.name])
-
-    bot_ids = set(chain.from_iterable(prefixer.user_login_to_node.get(u, ()) for u in local_bots))
-    await sdb.execute(
-        insert(Team).values(
-            Team(name=Team.BOTS, owner_id=account, parent_id=root_team_id, members=sorted(bot_ids))
-            .create_defaults()
-            .explode(),
-        ),
-    )
-    return len(bot_ids)
+    async with sdb.connection() as sdb_conn:
+        async with sdb_conn.transaction():
+            old_team = await fetch_bots.team(account, sdb_conn)
+            new_team = set(
+                chain.from_iterable(
+                    prefixer.user_login_to_node.get(u, ())
+                    for u in (bots - await fetch_bots.extra(mdb))
+                ),
+            )
+            if old_team is not None and set(old_team).issuperset(new_team):
+                return len(old_team)
+            log.info("writing %d bots for account %d", len(new_team), account)
+            sql = (await dialect_specific_insert(sdb))(Team)
+            if await is_postgresql(sdb):
+                sql = sql.on_conflict_do_update(
+                    constraint=Team.__table__.primary_key,
+                    set_={
+                        col.name: getattr(sql.excluded, col.name)
+                        for col in (Team.members, Team.updated_at)
+                    },
+                )
+            else:
+                sql = sql.prefix_with("OR REPLACE")
```

Comment:
on_conflict_do_update should work also with sqlite, see `assign_team_goals`

Reply:
Jeez, we have quite a few similar places in the existing code. Now I understand why you factored the dialect-specific insert but left the rest.